core/src/main/scala/org/apache/spark/sql/delta/DeltaOptions.scala

```diff
@@ -93,6 +94,17 @@ trait DeltaWriteOptionsImpl extends DeltaOptionParser {
     options.get(DATA_CHANGE_OPTION).exists(!toBoolean(_, DATA_CHANGE_OPTION))
   }
 
+  /** Whether to only overwrite partitions that have data written into it at runtime. */
+  val isDynamicPartitionOverwriteMode: Boolean = {
```

Comment:
It seems to me a lot of the `DeltaReadOptions` are written very similar to this. Not sure if they're treated differently. 

But our main reason for making this `val` instead of `def` was so that we would catch invalid values whenever `DeltaWriteOptions` is instantiated. If you think it's fine for that check to wait for `isDynamicPartitionOverwriteMode` to be called I'm okay with that.

Reply:
The `DeltaReadOptions` do not combine options with sql conf. So in terms of correctness, def vs val makes no difference. but in this case the presence sql conf makes it unique. and in that case, I would rather have same consistent behavior as the only the option that has the same unique property of relying on sql conf, mergeSchema.